src/storage/admin/AdminTaskManager.cpp

```diff
@@ -0,0 +1,171 @@
+/* Copyright (c) 2019 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "base/Base.h"
+#include "storage/admin/AdminTask.h"
+#include "storage/admin/AdminTaskManager.h"
+
+DEFINE_uint32(max_task_concurrency, 10, "The tasks number could be invoked simultaneously");
+DEFINE_uint32(max_concurrent_subtasks, 10, "The sub tasks could be invoked simultaneously");
+
+namespace nebula {
+namespace storage {
+
+using ResultCode = nebula::kvstore::ResultCode;
+using TaskHandle = std::pair<int, int>;     // jobid + taskid
+
+bool AdminTaskManager::init() {
+    LOG(INFO) << "max concurrenct subtasks: " << FLAGS_max_concurrent_subtasks;
+    pool_ = std::make_unique<nebula::thread::GenericThreadPool>();
+    CHECK(pool_->start(FLAGS_max_concurrent_subtasks));
+
+    bgThread_ = std::make_unique<thread::GenericWorker>();
+    CHECK(bgThread_->start());
+    bgThread_->addTask(&AdminTaskManager::pickTaskThread, this);
+
+    shutdown_ = false;
+    LOG(INFO) << "exit AdminTaskManager::init()";
+    return true;
+}
+
+void AdminTaskManager::addAsyncTask(std::shared_ptr<AdminTask> task) {
+    TaskHandle handle = std::make_pair(task->getJobId(), task->getTaskId());
+    LOG(INFO) << folly::stringPrintf("enqueue task(%d, %d), con req=%zu",
+                                     task->getJobId(), task->getTaskId(),
+                                     task->getConcurrentReq());
+    auto ctx = std::make_shared<TaskExecContext>(task);
+    tasks_.insert(handle, ctx);
+    taskHandleQueue_.add(handle);
+}
+
+nebula::kvstore::ResultCode
+AdminTaskManager::cancelTask(int jobId) {
+    LOG(INFO) << "AdminTaskManager::cancelTask() jobId=" << jobId;
+    auto ret = ResultCode::ERR_KEY_NOT_FOUND;
+    auto it = tasks_.begin();
```

Comment:
I suggest you just label the related task to be INVALID in taskContainer,  so you don't need to iterate the queue.  Because it is not thread safe.